module_3/gifts/service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,340 @@
+package com.epam.esm.service.impl;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.MultiValueMap;
+
+import com.epam.esm.dto.CertificateDto;
+import com.epam.esm.dto.TagDto;
+import com.epam.esm.exception.ErrorCode;
+import com.epam.esm.exception.NotFoundException;
+import com.epam.esm.exception.ValidationException;
+import com.epam.esm.repository.CertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.repository.model.CertificateModel;
+import com.epam.esm.repository.model.EntityConstant;
+import com.epam.esm.repository.model.TagModel;
+import com.epam.esm.service.CertificateService;
+import com.epam.esm.service.converter.CertificateConverter;
+import com.epam.esm.service.converter.TagConverter;
+import com.epam.esm.service.validation.CertificateValidation;
+import com.epam.esm.service.validation.TagValidation;
+import com.epam.esm.service.validation.Util;
+
+/**
+ * 
+ * Contains methods implementation for working mostly with
+ * {@code CertificateDto} entity.
+ *
+ */
+@Service
+public class CertificateServiceImpl implements CertificateService {
+	private static final int OFFSET = 0;
```

Comment:
@LizavetaSinitsyna It can be moved in separate class with all service const. DRY.

Reply:
@AliaksandrLiulis, I saw that there is repeat of constants (the same offset and limit), but I thought that maybe there can be cases when there will be decided to change this parameters only for the certificates for example. However I will move them as for now they are equal for all service classes.